packages/ra-ui-materialui/src/field/ReferenceField.tsx

```diff
@@ -103,54 +109,48 @@ ReferenceField.defaultProps = {
     link: 'edit',
 };
 
-export interface ReferenceFieldProps<RecordType extends RaRecord = any>
-    extends PublicFieldProps,
-        InjectedFieldProps<RecordType> {
+export interface ReferenceFieldProps<
+    RecordType extends Record<string, unknown> = Record<string, any>,
+    ReferenceRecordType extends RaRecord = RaRecord
+> extends Omit<FieldProps<RecordType>, 'source'>,
+        Required<Pick<FieldProps<RecordType>, 'source'>> {
     children?: ReactNode;
-    queryOptions?: UseQueryOptions<RecordType[], Error> & { meta?: any };
+    queryOptions?: UseQueryOptions<ReferenceRecordType[], Error> & {
+        meta?: any;
+    };
     reference: string;
-    resource?: string;
-    source: string;
     translateChoice?: Function | boolean;
-    link?: LinkToType<RecordType>;
+    link?: LinkToType<ReferenceRecordType>;
     sx?: SxProps;
 }
 
 /**
  * This intermediate component is made necessary by the useReference hook,
  * which cannot be called conditionally when get(record, source) is empty.
  */
-export const NonEmptyReferenceField: FC<
-    Omit<ReferenceFieldProps, 'source'> & { id: Identifier }
-> = ({ children, id, record, reference, link, queryOptions, ...props }) => {
-    const createPath = useCreatePath();
-    const resourceDefinition = useResourceDefinition({ resource: reference });
-
-    const resourceLinkPath =
-        link === false ||
-        (link === 'edit' && !resourceDefinition.hasEdit) ||
-        (link === 'show' && !resourceDefinition.hasShow)
-            ? false
-            : createPath({
-                  resource: reference,
```

Comment:
Any particular reason you moved that logic down to PureReferenceFieldView? By the way that's a BC, as ReferenceFieldView is public.